haystack/schema.py

```diff
@@ -777,13 +795,35 @@ def _pydantic_dataclass_from_dict(dict: dict, pydantic_dataclass_type) -> Any:
     return dataclass_object
 
 
+def _dict_factory(list_of_tuples):
+    """Meant to be as the dict_factory for `asdict`. This function is called within `asdict` to convert a list of tuples
+    into a dictionary object. This handles the conversion of pandas Dataframes into a list of lists.
+
+    :param list_of_tuples: list of (key, value) pairs
+    """
+    res = {}
+    for key, val in list_of_tuples:
+        if isinstance(val, pd.DataFrame):
+            val = dataframe_to_list(val)
+        res.update({key: val})
+    return res
+
+
 class NumpyEncoder(json.JSONEncoder):
     def default(self, obj):
         if isinstance(obj, np.ndarray):
             return obj.tolist()
         return json.JSONEncoder.default(self, obj)
 
 
+def dataframe_to_list(df: pd.DataFrame) -> List[List]:
+    return [df.columns.tolist()] + df.values.tolist()
+
+
+def dataframe_from_list(list_df: List[List]) -> pd.DataFrame:
+    return pd.DataFrame(columns=list_df[0], data=list_df[1:])
```

Comment:
Do we really need these methods? 
They're one liners that don't do much, we might as well keep the inline.

Reply:
It was more to centralize the place of how we convert dataframes to a json serializable object. So hopefully in the future if we need to do this in other places in the code we won't need to reimplement the one-liner or remember which one-liner to use. 

But I see your point, people could also forget to use this function and just reimplement the one-liner (even an incorrect one) so it may not be worth it.

Reply:
Hey I'm leaning towards leaving this as is unless you think it's worth removing. What are your thoughts?

Reply:
Good for me if we want to leave them.